### ISSUE TITLE
ALSA/Pulseaudio crackling

### ISSUE THREAD

**Issue by tyler274** [RPCS3/rpcs3#3087]
Affects P5 but also drakengard 3 and project diva. 

Can be temporarily fixed by unplugging my speakers and plugging them back in, after a period of intense crackling it goes away for a short time (a few minutes at least.) 

probably partly pulse's fault but only rpcs3 is having this issue for me.

**Comment by aFoxNamedMorris** [comment]
I got around this by switching to OpanAL, of all things.

**Comment by petran79** [comment]
Pulseaudio option does not appear in my case. Just ALSA and OpenAL. Switching to OpenAL causes slowdown and makes crackling worse

**Comment by tyler274** [comment]
@petran79 puleaudio using alsa as its backend

**Comment by greentop** [comment]
Does starting RPCS3 in terminal with PULSE_LATENCY_MSEC=30 change anything?

:~$ PULSE_LATENCY_MSEC=30 '/home/YourHomeNameHere/rpcs3/bin/rpcs3'

**Comment by tyler274** [comment]
@greentop since I followed https://wiki.archlinux.org/index.php/Professional_audio and setup Pulse->Audio->Jack for audio and got my reported audio latency down to ~5ms its far less prominent.
Setting the pulse latency like that makes it more pronounced, but right now its still not as bad as before (without changing kernel mods and such to verify.)

**Comment by unknown321** [comment]
What audiocard do you use? I have horrible cracking on my usb headset, but there are little to none problems if main output is built-in card.
Pulse log, built-in card:
```
( 224.411|   2.015) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 224.411|   0.000) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 224.506|   0.095) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 225.113|   0.606) D: [alsa-source-USB Audio] memblock.c: Pool full
( 225.131|   0.017) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 225.173|   0.042) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 225.273|   0.100) D: [alsa-source-USB Audio] memblock.c: Pool full
( 225.274|   0.001) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 225.333|   0.058) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 225.456|   0.123) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 225.462|   0.005) D: [pulseaudio] memblock.c: Pool full
( 231.451|   5.989) D: [alsa-sink-ALC887-VD Analog] ratelimit.c: 33 events suppressed
( 231.451|   0.000) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
( 231.458|   0.007) D: [alsa-source-USB Audio] memblock.c: Pool full
( 231.461|   0.003) D: [pulseaudio] memblock.c: Pool full
( 231.701|   0.239) D: [alsa-sink-ALC887-VD Analog] memblock.c: Pool full
```
Pulse log, Plantronics GameCom 780 usb headset, a lot of cracking:
```
(  14.836|   0.000) D: [alsa-sink-USB Audio] source.c: Processing rewind...
(  14.836|   0.000) D: [alsa-sink-USB Audio] protocol-native.c: Implicit underrun of 'ALSA Playback'
(  14.836|   0.000) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 160 bytes ago (1280 bytes ahead in playback buffer)
(  14.836|   0.000) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 160 bytes ago (1280 bytes ahead in playback buffer)
(  14.836|   0.000) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 160 bytes ago (1280 bytes ahead in playback buffer)
(  14.836|   0.000) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 160 bytes ago (1280 bytes ahead in playback buffer)
(  14.836|   0.000) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 160 bytes ago (1280 bytes ahead in playback buffer)
(  14.840|   0.003) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 1312 bytes ago (128 bytes ahead in playback buffer)
(  14.840|   0.000) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 1312 bytes ago (128 bytes ahead in playback buffer)
(  14.841|   0.000) D: [pulseaudio] module-suspend-on-idle.c: Sink alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo becomes idle, timeout in 5 seconds.
(  14.841|   0.000) D: [alsa-sink-USB Audio] sink.c: alsa_output.usb-Plantronics_Plantronics_GameCom_780-00.analog-stereo: Found underrun 1312 bytes ago (128 bytes ahead in playback buffer)
(  14.841|   0.000) D: [alsa-sink-USB Audio] alsa-sink.c: hwbuf_unused=0
(  14.841|   0.000) D: [alsa-sink-USB Audio] alsa-sink.c: setting avail_min=88009
(  14.841|   0.000) D: [alsa-sink-USB Audio] alsa-sink.c: Requested to rewind 352800 bytes.
(  14.841|   0.000) D: [alsa-sink-USB Audio] alsa-sink.c: Limited to 1184 bytes.
(  14.841|   0.000) D: [alsa-sink-USB Audio] alsa-sink.c: before: 296
(  14.841|   0.000) D: [alsa-sink-USB Audio] alsa-sink.c: after: 296
(  14.841|   0.000) D: [alsa-sink-USB Audio] alsa-sink.c: Rewound 1184 bytes.
(  14.841|   0.000) D: [alsa-sink-USB Audio] sink.c: Processing rewind...
(  14.841|   0.000) D: [alsa-sink-USB Audio] source.c: Processing rewind...
```

**Comment by tyler274** [comment]
im just using the internal intel one. when i used my usb DAC it was also extremely bad.

**Comment by C0rn3j** [comment]
Also have this issue with my HyperX cloud 2 headset and speakers plugged into the case's audio jack.

**Comment by AccountOneOff** [comment]
Still present in v0.0.3-43b860d-7866_linux64.AppImage [2017-08-29] even with Pulse option.

**Comment by C0rn3j** [comment]
I think the crackling only happens during slowdowns for me when I use the PulseAudio option.

**Comment by Illynir** [comment]
Same problem here. Very boring.

**Comment by Nekotekina** [comment]
I would try ALSA with different values for Buffer Count setting in config.yml, however, I doubt it helps. About 5 ms latency seems pretty much expected.

**Comment by Illynir** [comment]
I have tried many values for buffer count, no difference.

I also tried a lot of different Pulseaudio configurations, without success even if it is sometimes slightly better. Note that it doesn't work on all games (On Nier for example it's catastrophic) and that it doesn't depend on the speed of the game (fullspeed or not) even if it's worse when you slow down.

Please note that I have no problems except on RPCS3.

Why not think about going through Portaudio with SDL?

**Comment by kd-11** [comment]
Pause emulation for 5 seconds then resume. Pulseaudio gets magically fixed like that. Its something to do with sync imo.

**Comment by Illynir** [comment]
I did some tests that I recorded to show you the problem. I did the test on Nier. To avoid any problem with the slowdowns of the game I went to the menu where I am fullspeed 60 FPS.

Nier with Pulseaudio:
https://clyp.it/fpkeswk0

Nier with Pulseaudio  Pause/Resume 10 seconds:
https://clyp.it/3c2x0xcc

Not really of difference for me, maybe slightly better...

I can't manage to record Alsa and OpenAL directly for the moment but know that on Alsa it's the same as Pulseaudio and OpenAL is much worse.

On Windows I don't have this problem.

**Comment by Illynir** [comment]
I've done a lot of testing again, try all possible and imaginable combinations of configuration and the problem persists. For example on Naruto where I am fullspeed all the time the sound is catastrophic whether it is Alsa, Pulseaudio or OpenAL.

I don't get much feedback from people on Linux and when I ask for some tests I never get anything... so it's hard to know the cause but for me it's clearly a buffer/latency problem.

Apart from RPCS3 the sound works perfectly everywhere else including other emulators such as: Dolphin, PPSSPP, PCSX2, Retroarch, native linux, browser, video, etc....

**Comment by Illynir** [comment]
Problem is visible on this video for example: https://www.youtube.com/watch?v=WC0ESMa-Ips

**Comment by Illynir** [comment]
I record the sound on Naruto: https://clyp.it/n20p4oel

Logo intro + cinematic + main menu.

**Comment by Nekotekina** [comment]
This may be unrelated, but may be worth trying to disable speedstep (put all CPU cores to their nominal frequency) on Linux:
```sudo -s
for CPUFREQ in /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor; do [ -f $CPUFREQ ] || continue; echo -n performance > $CPUFREQ; done
exit```

This fixed some problems for me.

**Comment by Illynir** [comment]
I already had the speedstep disabled since my processor is OC. But after a lot of tests and confirmation by Kd I think the general problem is that on Linux the priority of threads is poorly managed or not existing on RPCS3. I guess it's kernel scheduling that takes the lead, so normally audio threads that should be high priority are not.

It works more or less when there are enough threads/core available but when you switch to an i5 with only 4 threads it is problematic on some games.

This also explains some performance improvements obtained on Linux and not on windows.

**Comment by Malbeth81** [comment]
I'm having the same issue across several games!

OS: Arch Linux x86_64
Host: GB-BNi7HG6-1060 1.x
Kernel: 4.16.3-1-ARCH
Resolution: 1920x1080
CPU: Intel i7-7700HQ (8) @ 3.800GHz
GPU: NVIDIA GeForce GTX 1060 Mobile 3GB
Memory: 2238MiB / 7916MiB

**Comment by ZahnPastor** [comment]
yeah, i noticed that when the game goes below 29FPS, which it normally runs on like 29,96, that the audio is basically completely missing, with the ```PULSE_LATENCY_MSEC=30``` environment variable (The game is Drakengard 3 btw)

**Comment by Nekotekina** [comment]
Does anyone have other programs playing sound in parallel when the problem happens?

**Comment by ZahnPastor** [comment]
the audio options didnt show me any other applications that made sound while i just tested now, so no

**Comment by Malbeth81** [comment]
For me I'm only using Alsa. Recently I noticed similar issues with other games so it's possible its caused by a recent package update. However Rpcs3 showed the issue weeks before, when the other games were fine.

---- On Wed, 30 May 2018 15:23:54 -0400 notifications@github.com wrote ----

the audio options didnt show me any other applications that made sound while i just tested now, so no

**Comment by AniLeo** [comment]
Is this still an issue?

**Comment by z0z0z** [comment]
@AniLeo 
Just built RPCS3 from git.

ALSA backend gives me no audio output, but it shows up in pavucontrol. Other ALSA programs work.

Pulse has crackling.

OpenAL seems to work best but I seem to need to disable buffering for some games (At least Project Diva F 2nd).

**Comment by Nekotekina** [comment]
@z0z0z Is something in console with ALSA?

Also I'd suggest to try to disable tsched like here: https://askubuntu.com/questions/371595/for-pulseaudio-what-does-tsched-do-and-what-are-the-defaults

**Comment by z0z0z** [comment]
@Nekotekina 
Sorry I am actually getting ALSA output now. I don't know what that was about.

Pulse still crackling in at least some games though, Nier for example.

**Comment by RainbowCookie32** [comment]
Still an issue in current master. If you leave a game running for a while, eventually audio breaks and stays broken if you restart the game or RPCS3. 
Used Nier and ALSA for testing, same issue on PulseAudio.
https://streamable.com/652ne2

**Comment by nitroos-ai** [comment]
Try this #11081.

**Comment by Darkhost1999** [comment]
So from what I read the original issue was about crackling and popping. Then I read about audio never recovering. And then again about audio never recovering after you leave for some time.

I can tell you if the issue is not about bad audio but about never recovering. Than the issue is not isolated to Linux.

On my windows 11 and was happening on 10. Since the moment I very first downloaded RPCS3 and all of this time. If you go to go get a coffee, lunch, or leave to go to the gas station and take longer than a half an hour every single game I own in RPCS3 will lose it's audio and never recover.

No the new PR does not fix this issue. I am confident in the issue having nothing to do with audio backends depending on whatever this mess is pertaining to. My recommendation is that this issue is closed. And a separate ticket is created due to the age of the issue. And upcoming changes to the audio backends requiring new information to be provided.

**Comment by Vestrel** [comment]
Does it happen in buffered/unbuffered mode?

**Comment by Darkhost1999** [comment]
I'll edit this message as I acquire logs to make it easier for me to focus and to provide information.

I'm testing all on Cubeb:
- With buffer and OS scheduler

- With buffer 5ms and OS scheduler

- Without buffer and OS scheduler

- With buffer and RPCS3 scheduler

- With buffer 5ms and RPCS3 scheduler

- Without buffer and RPCS3 scheduler

- With buffer and RPCS3 Alt scheduler

- With buffer 5 ms and RPCS3 Alt scheduler

- Without buffer and RPCS3 Alt scheduler

Where I'm using windows 11 with an AMD Ryzen 5 3600 GTX 1660S on an Omen mobo 87c3.

I've already tested all these options on XAudio2 through the course of me using RPCS3 and testing such games as Red Dead Redemption and which scheduler worked best. And trying to find out why the audio thread dies. Each test takes about 2 hours to perform. 1st the controller needs to disconnect via timeout. Then you just gotta let it sit there for about 1 to 2 hours while you do other things to come back to RPCS3 reconnect the controller and find that audio is just gone and playing for 5 minutes vs 1 hour makes no difference until the game is restarted or the emulator is rebooted.

**Comment by Darkhost1999** [comment]
I am not certain as to what changed. Other than me updating to the PRs latest build today vs 2 days ago. The PR named audio.exe has been running the emulator since 9am. It's now 11:30am and audio has not died yet using cubeb.
Audio has always died in this amount of time for XAudio2.

Steps to create the issue are normally just play a game long enough to get in game. Then just walk away. Leave your controller on and everything. Let the controller turn off by timeout. Then 2 hours later or so come back and reconnect the controller. Press X and realize no audio is present.

I'll wait longer to see what happens.

**Comment by orbea** [comment]
@nitroos-ai, @Vestrel I tried PR https://github.com/RPCS3/rpcs3/pull/11081 and it did not improve the situation. I think the solution is to use a resampler like speex, potentially cubeb could do this, but I don't think its hooked up for a lot of audio backends like alsa or pulseaudio.

**Comment by nitroos-ai** [comment]
A resampler has been added in  #11358.
Checking in Audio:"Enable Time Stretching" fixed most cracklings in games for me.

**Comment by RainbowCookie32** [comment]
The issue here wasn't just crackling because of low performance, but these 2 specific backends getting broken audio after playing for a while. Time Stretching just ""hides"" the issue anyways, not really a fix

**Comment by Vestrel** [comment]
If time stretching fixes crackling then the problem either in cellAudio or in game code.

**Comment by Megamouse** [comment]
Or potato PCs

**Comment by Asinin3** [comment]
Alsa/Pulseaudio aren't available in RPCS3 for awhile now. Closing.
